lib/identity_cache/cache_fetcher.rb

```diff
@@ -24,27 +75,181 @@ def fetch_multi(keys, &block)
       results
     end
 
-    def fetch(key)
-      result = nil
-      yielded = false
-      @cache_backend.cas(key) do |value|
-        yielded = true
-        unless IdentityCache::DELETED == value
-          result = value
-          break
+    def fetch(key, **cache_options)
+      if use_fill_lock?(cache_options)
+        fetch_with_fill_lock(key, cache_options) { yield }
+      else
+        fetch_without_fill_lock(key) { yield }
+      end
+    end
+
+    private
+
+    def use_fill_lock?(fill_lock_duration: nil, lock_wait_limit: nil)
+      fill_lock_duration && IdentityCache.should_fill_cache?
+    end
+
+    def fetch_without_fill_lock(key)
+      data = nil
+      cas_or_add(key) do |value|
+        value = nil if value == IdentityCache::DELETED || FillLock.cache_value?(value)
+        unless value.nil?
+          return value
         end
-        result = yield
+        data = yield
         break unless IdentityCache.should_fill_cache?
-        result
+        data
+      end
+      data
+    end
+
+    def fetch_with_fill_lock(key, fill_lock_duration:, lock_wait_limit: 2)
+      raise ArgumentError, 'fill_lock_duration must be greater than 0.0' unless fill_lock_duration > 0.0
+      raise ArgumentError, 'lock_wait_limit must be greater than 0' unless lock_wait_limit > 0
+      lock = nil
+      using_fallback_key = false
+      expiration_options = EMPTY_HASH
+      (lock_wait_limit + 2).times do
```

Comment:
I'm not sure what the `+2` is doing, do you mind explaining the reason for this? 🙏

Reply:
1 for the initial iteration before it could lock wait and 1 is for the switch to using the fallback key after which it needs to try again without waiting.  Perhaps an in code comment would help clarify this.

This line is mostly just to catch unexpected infinite loops, so should be replaceable with `loop do`.